aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -110,55 +141,149 @@ impl<V: TransactionWrite> DataRead<V> {
         let self_kind = self.get_kind();
         let other_kind = other.get_kind();
 
-        if self_kind < other_kind {
-            DataReadComparison::Insufficient
-        } else {
-            let downcast_eq = if self_kind == other_kind {
-                // Optimization to avoid unnecessary clones (e.g. during validation).
-                self == other
-            } else {
-                self.downcast(other_kind)
-                    .expect("Downcast to lower kind must succeed")
-                    == *other
-            };
-
-            if downcast_eq {
+        if self_kind == other_kind {
+            // Optimization to avoid unnecessary clone in convert_to. Has been optimized
+            // because contains is called during validation.
+            if self == other {
                 DataReadComparison::Contains
             } else {
                 DataReadComparison::Inconsistent
             }
+        } else {
+            match self.convert_to(&other_kind) {
+                Some(value) => {
+                    if value == *other {
```

Comment:
Now does this dereference work, do we copy? `&value == other` should also be ok?

Reply:
good point. It's small types, but still.